src/modules/user/core/services/kyc.service.new.ts

```
@@ -0,0 +1,190 @@
+    if (!kycData || !kycData.documents || kycData.documents.length === 0) {
+      throw new MissingKycDataError();
+    }
+
+    const { country, subtype, type, fields } = kycData.documents[0];
+    const documentNumber = fields.documentNumber.value;
+    const curp = this.constants.METAMAP_SUBTYPES.MX.includes(subtype) ? fields.curp.value : null;
+    const address = this.constants.METAMAP_SUBTYPES.MX.includes(subtype) ? fields.address.value : null;
+
+    currentKyc.country = country;
+    currentKyc.address = address;
+    currentKyc.curp = curp;
+    currentKyc.subtype = subtype ?? type;
+    currentKyc.documentNumber = documentNumber;
+
+    await this.kycRepository.update(currentKyc);
+  }
+
+  async saveKycData(uid: string) {
+    const userIdEntity = EntityId.fromString(uid);
+    const currentKyc = await this.kycRepository.getByUserId(userIdEntity);
+
+    if (!currentKyc) {
+      await this.createKycData(uid);
+    } else {
+      await this.updateKycData(uid, currentKyc);
+    }
+  }
+
+  private determineCurrentStatus(status: string): string {
+    switch (status) {
+      case MetamapStatus.REVIEW_NEEDED:
+        return MetamapStatus.REVIEW_NEEDED;
+      case MetamapStatus.REJECTED:
+        return MetamapStatus.REJECTED;
+      default:
+        return status.toUpperCase();
+    }
+  }
+
+  async updateUserVerified(uid: string, status: string) {
+    const userIdEntity = EntityId.fromString(uid);
+    const user = await this.userRepository.getById(userIdEntity);
+
+    const currentStatus = this.determineCurrentStatus(status);
+
+    user.kycStatus = currentStatus;
+    await this.userRepository.update(user);
+
+    await this.saveKycData(uid);
+    // TODO: notification service
+  }
+
+  async cancelKycStatus(uid: string): Promise<boolean> {
+    const userIdEntity = EntityId.fromString(uid);
+    const user = await this.userRepository.getById(userIdEntity);
+
+    if (user.kycStatus !== 'VERIFIED') {
+      user.kycStatus = null;
```

Wouldn't it be better to have an unverified status for the user instead of setting it to null? Also I think we should throw an error if this method is called and the user is already verified